source/extensions/io_socket/user_space/io_socket_handle_impl.cc

```diff
@@ -0,0 +1,330 @@
+  }
+  // Closed peer.
+  if (!peer_handle_) {
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // Error: write after close.
+  if (peer_handle_->isPeerShutDownWrite()) {
+    // TODO(lambdai): `EPIPE` or `ENOTCONN`.
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // The peer is valid but temporary not accepts new data. Likely due to flow control.
+  if (!peer_handle_->isWritable()) {
+    return {0, Api::IoErrorPtr(Network::IoSocketError::getIoSocketEagainInstance(),
+                               Network::IoSocketError::deleteIoError)};
+  }
+
+  auto* const dest_buffer = peer_handle_->getWriteBuffer();
+  // Write along with iteration. Buffer guarantee the fragment is always append-able.
+  uint64_t bytes_written = 0;
+  for (uint64_t i = 0; i < num_slice && !dest_buffer->highWatermarkTriggered(); i++) {
+    if (slices[i].mem_ != nullptr && slices[i].len_ != 0) {
+      dest_buffer->add(slices[i].mem_, slices[i].len_);
+      bytes_written += slices[i].len_;
+    }
+  }
+  peer_handle_->setNewDataAvailable();
+  ENVOY_LOG(trace, "socket {} writev {} bytes", static_cast<void*>(this), bytes_written);
+  return {bytes_written, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+}
+
+Api::IoCallUint64Result IoSocketHandleImpl::write(Buffer::Instance& buffer) {
+  // Empty input is allowed even though the peer is shutdown.
+  if (buffer.length() == 0) {
+    return Api::ioCallUint64ResultNoError();
+  }
+  if (!isOpen()) {
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // Closed peer.
+  if (!peer_handle_) {
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // Error: write after close.
+  if (peer_handle_->isPeerShutDownWrite()) {
+    // TODO(lambdai): `EPIPE` or `ENOTCONN`.
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // The peer is valid but temporary not accepts new data. Likely due to flow control.
+  if (!peer_handle_->isWritable()) {
+    return {0, Api::IoErrorPtr(Network::IoSocketError::getIoSocketEagainInstance(),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  uint64_t total_bytes_to_write = 0;
+  const uint64_t max_bytes_to_write = buffer.length();
+  while (peer_handle_->isWritable()) {
+    const auto& front_slice = buffer.frontSlice();
+    if (front_slice.len_ == 0) {
+      break;
+    } else {
+      peer_handle_->getWriteBuffer()->move(buffer, front_slice.len_);
+      total_bytes_to_write += front_slice.len_;
```

Comment:
You need to flip the previous 2 lines.

front_slice may not be valid after the move due to copy of small slices on move optimization.  The current version can result in use-after-free, see coalesceOrAddSlice